### PR TITLE
Aggregator smaller model outputs  - Ready to be reviewed

### PR DESCRIPTION
**Pull request by wahabk**
This is to reference #{1001}.

I have described the issue and planned fix in detail I believe,  this is a draft PR to ask for some help

I have made most changes I believe, but this is a broken draft to show you my idea, let me know if the assertions are insufficient or if argument/attribute names are not clear

The bit I'm confused with is the subject padding, the subject is now forced to be padded and the padding is set to be at least the difference between input and output. and the final aggregator output tensor isn't padded 

So say I have an input tensor with shape `(128.128.128)` , with a patch_size  of `(64,64,64)`  and a model output size of `(60,60,60)`

The patch_difference is set to be `(4,4,4)` and overlap is made to be at least equal to this, when the subject is padded in the `GridSampler` does that mean it's shape is now `(136,136,136)`?

**Description**
https://github.com/fepegar/torchio/issues/1001

**Checklist**

- [x] I have read the [`CONTRIBUTING`](https://github.com/fepegar/torchio/blob/main/CONTRIBUTING.rst) docs and have a developer setup (especially important are `pre-commit`and `pytest`)
- [x] Non-breaking change (would not break existing functionality)
- [ ] Breaking change (would cause existing functionality to change)
- [x] Tests added or modified to cover the changes
- [x] Integration tests passed locally by running `pytest`
- [x] In-line docstrings updated
- [x] Documentation updated, tested running `make html` inside the `docs/` folder
- [x] This pull request is ready to be reviewed
- [x] If the PR is ready and there are multiple commits, I have [squashed them and force-pushed](https://www.w3docs.com/snippets/git/how-to-combine-multiple-commits-into-one-with-3-steps.html#force-pushing-commits-7)

### CODE REVIEW

**Comment by wahabk** [comment]
This is to reference #{1001}.

I have described the issue and in detail I believe https://github.com/fepegar/torchio/issues/1001, I think this is PR is ready to be merged except for the final questions. I'm  still getting my feet wet with PRs  so would  appreciate  some guidance.

TLDR for why this is important:

I believe tio grid sampling and  aggregating should be able to handle smaller model outputs. My model predictions are terrible even with averaging or hann windowing. Unfortunately most popular model libraries (such as the great monai) only provide models with the same output size and input. But it is crucial in my application to let the model see a bigger input ROI than semantic label outputs - by **padding convolutions**, as this gives _context_ for the prediction. The original unet paper uses padded convolutions for smaller outputs than inputs.

**Description**

**Summary of changes:**
- Add argument  to `GridSampler` called `model_output_size` in case it is smaller than `patch_size`
- This then computes `patch_diffs` which is the difference between them
- `GridSampler` now makes  sure  that overlap_size is at least the same as `patch_diffs`
- The locations are changed to fit the smaller patch_size
- If overlap_mode is 'crop' it changes `GridAggregator._crop_patch`  to only crop  `half_overlap - patch_diffs`
- If overlap_mode is 'hann' it makes the hann window of shape `model_output_size` instead of `patch_size`

**Final questions before you agree:**
- The reason this bug was hidden from me is because the assertion in `GridAggregator.add_batch`  checks `target_spatial_shape` from the provided batch, and _not from self.patch_size_.  I belive this should be changed?
- Does the kwarg name `model_output_size` make sense?
- Are my new assertions thorough  enough? Particularly  in `GridSampler.__init__()` and `GridAggregator.add_batch()`
- I have added a for loop in `test_inference.py` and  `test_inference_smaller.py`  that checks for different overlap_modes, however, the default tests fail under overlap mode 'hann'

**Checklist**




- [x] I have read the [`CONTRIBUTING`](https://github.com/fepegar/torchio/blob/main/CONTRIBUTING.rst) docs and have a developer setup (especially important are `pre-commit`and `pytest`)
- [x] Non-breaking change (would not break existing functionality)
- [ ] Breaking change (would cause existing functionality to change)
- [x] Tests added or modified to cover the changes
- [x] Integration tests passed locally by running `pytest`
- [x] In-line docstrings updated
- [x] Documentation updated, tested running `make html` inside the `docs/` folder
- [x] This pull request is ready to be reviewed
- [x] If the PR is ready and there are multiple commits, I have [squashed them and force-pushed](https://www.w3docs.com/snippets/git/how-to-combine-multiple-commits-into-one-with-3-steps.html#force-pushing-commits-7)

**Comment by wahabk** [comment]
@fepegar This is ready now when you have time to review. Please read the second comment, the first was the draft PR.

I tried squashing my commits using the tutorial, did it work correctly?

Any edits or requests you have in mind let me know and I will edit the PR.

**Comment by fepegar** [comment]
Hi, @wahabk. Thanks for the PR. Apologies for the delay. I will take a look at all this as soon as possible.

File: src/torchio/data/inference/aggregator.py [hunk 1]
```diff
@@ -143,8 +163,20 @@ def add_batch(
         # There should be only one patch size
         assert len(np.unique(patch_sizes, axis=0)) == 1
         input_spatial_shape = tuple(batch.shape[-3:])
-        target_spatial_shape = tuple(patch_sizes[0])
-        if input_spatial_shape != target_spatial_shape:
+        # Should target size be self.patch_size?
+        target_spatial_shape = tuple(
```
Comment (fepegar) [inline src/torchio/data/inference/aggregator.py:167]: Why is it a tuple with only one element?

**Comment by fepegar** [comment]
Closing this as the review was ignored. Thanks anyway for your contribution. Feel free to reopen if needed.
